### ISSUE TITLE
Intensity quantization and method?

### ISSUE THREAD

**Issue by kdk411** [Radiomics/pyradiomics#378]
Hi,

I looked through issues thread and settings examples, and if I am not mistaken, binwidth is the parameter for intensity quantization. And the default value is 25.

So say, I have pre-processed (like N4 corrected, then normalized by subtracting the mean intensity of normal region, then divided by std, etc) the image intensity such that values range from -4 to 4 in double precision. And I didn't pass the binwidth parameter as my settings for feature extractor.

In this scenario, then every voxel intensity is now assigned to... discrete values between 0 and 25...? (or 24?) Also, how is it assigned to...? equal...? Lloyd...? Is there an option for such method?

**Comment by JoostJM** [comment]
[This](http://pyradiomics.readthedocs.io/en/latest/radiomics.html#radiomics.imageoperations.binImage) section of the documentation details how the binWidth parameter is used.

tldr: It discretizes the grayvalues into equally spaced bins of <binWidth> grayvalues, number of discreet grayvalues present is then dependent on the range in the ROI.

In your case, binwidth 25 will lead to flat regions for texture, I'd advise a binwidth of < 1.

**Comment by kdk411** [comment]
Thank you for your link!

And yes, you have a great point. If I use 25 as my binwidth, then it is probably too big such that every voxel would fall into one bin...

**Comment by kdk411** [comment]
Sorry, I thought I was good to go, but I must ask this before moving on.

So I checked my logger and saw the following:

```
DEBUG:radiomics.imageoperations: Calculated 24 bins for bin width 0.25 with edges: [-4.   -3.75 -3.5  -3.25 -3.   -2.75 -2.5  -2.25 -2.   -1.75 -1.5  -1.25
 -1.   -0.75 -0.5  -0.25  0.    0.25  0.5   0.75  1.    1.25  1.5   1.75
  2.  ])
DEBUG:radiomics.glcm: Calculating GLCM matrix in C
DEBUG:radiomics.glcm: Process calculated matrix
DEBUG:radiomics.glcm: Create symmetrical matrix
DEBUG:radiomics.glcm: No empty angles
DEBUG:radiomics.glcm: Calculating GLCM coefficients
DEBUG:radiomics.glcm: GLCM feature class initialized, calculated GLCM with shape (7, 7, 4)
DEBUG:radiomics.glcm: Calculating features
```
Now, for this particular patient data, I know that the intensity ranges from -4 to 2, so bin edges make sense. Also, based on what is described under `imageoperations.binImage()`, image intensities are mapped to 1,2,3, .., 7, so GLCM of shape (7,7,4) also makes sense (and I am doing in 2D, so 4 checks out).

But here is more of a philosophical (?) question. I thought the whole point of discretizing intensities was to putting intensity range for different patients into the same range, such that we are comparing apples to apples, not apples to oranges...

For instance, let's say patient A has intensity range from 50 ~ 200, and patient B has intensity range from 30~ 400. Then after discretizing the intensity values into a fixed range, like 64, for instance, then both patient A and B are now into the same intensity range.

But with the method described here, because we are specifying the binwidth, not the level to be discretized into, I don't think that can be achieved...? 

Or did I misunderstand the documentation...?

**Comment by kdk411** [comment]
does anyone have any thoughts on this matter? I am still confused on this issue and hopefully someone can at least throw 2 cents on this...

**Comment by fedorov** [comment]
@kdk411 sorry for lack of response on this one. 

In my view (and I don't have anywhere close as much expertise as @JoostJM on this matter!), you are indeed asking a philosophical question, since I do not think there is unequivocal evidence to suggest that one approach should always be used than the other (fixed bin width _vs_ fixed number of bins). There are also different approaches to image normalization. We discussed this in the past, and I think we reached the agreement that ideally _pyradiomics_ should provide the choice to the user between the two binning approaches, but I don't think this has been implemented. I also encourage you to review the work being done by the [Image Biomarkers Standardization Initiative (IBSI)](https://arxiv.org/abs/1612.07003), and also check out their newly set up [IBSI google group](https://groups.google.com/forum/#!forum/the_ibsi). I think that group could be a perfect place for philosophical questions like this one.

Note that this response is my personal impression/understanding, and may not agree with the views of other developers of _pyradiomics_.

**Comment by JoostJM** [comment]
@kdk411 Sorry, I missed your reopening of this issue.

On the topic of fixed bin width vs. fixed bin count I agree with @fedorov that we should implement the fixed bin count somewhere. However, we generally advise to use a fixed bin width, as the meaning of "contrast" (which is what you're looking at when calculating texture features) is dependent on the *range*  of gray values when you're using a fixed bin width, essentially causing you to compare apples to oranges (a contrast of say, 100 HU is a lot different than a contrast of 1 HU).

Using a fixed bin width means that you're always considering a similar contrast. If an image has more contrast, that it will cause you to have more discretized gray values and hence different features, but IMHO that is OK, as I consider that to be a different texture (enter the philosophical debate...).

On top of all that I believe that using a fixed bin width is more stable against outliers (consider a single large outlier: in a fixed bin count, this will cause the width to be very large, resulting in texture features that reflect a more homogeneous texture).

Still, all this assumes that you're original gray values have similar meaning, but if that is not the case, it is my view the images should be normalized, as otherwise your first order features would also be uncomparable (as they are largely based upon the input gray values, i.e. not discretized).

I tried to explain this as well in the [FAQ](http://pyradiomics.readthedocs.io/en/latest/faq.html#why-is-there-no-parameter-to-specify-a-fixed-bin-count), as it is one of the more common questions we receive.

As a closing remark, I would like to add that this is my opinion and certainly not absolute truth or anything (and contrary to the methods used in many published papers on Radiomics). For me, bin width vs. bin count really illustrates the pressing need for researchers in Radiomics to come together a bit more and try to reach consensus on how to do Radiomics (e.g. through initiatives such as IBSI).

**Comment by kdk411** [comment]
First of all, I really appreciate everyone who shared thoughts/ideas on this matter!

Yes, a lot of papers that I have been exposed to used fixed number of bins, hence I was biased towards that method with the idea of comparing "apples to apples". But now after seeing arguments from using "fixed binwidth", I can also appreciate their perspectives.

At the end, I can see both methods have pros and cons... and it is indeed something that radiomics community needs to come together and have a consensus on it.

And, another lesson to take is... I should read FAQ section more carefully. :+1:
